modules/client/src/client-common/utils.ts

```diff
@@ -121,3 +146,135 @@ export function votingModeFromContracts(votingMode: number): VotingMode {
       throw new InvalidVotingModeError();
   }
 }
+
+export function applyInstallatonParamsToContract(
+  params: ApplyInstallationParams,
+): PluginSetupProcessor.ApplyInstallationParamsStruct {
+  return {
+    plugin: params.pluginAddress,
+    pluginSetupRef: {
+      pluginSetupRepo: params.pluginRepo,
+      versionTag: params.versionTag,
+    },
+    helpersHash: keccak256(
+      defaultAbiCoder.encode(["address[]"], [params.helpers]),
+    ),
+    permissions: params.permissions.map((permission) => {
+      return { ...permission, condition: permission.condition || AddressZero };
+    }),
+  };
+}
+
+export function applyInstallatonParamsFromContract(
+  result: Result,
+): DecodedApplyInstallationParams {
+  const params = result[1];
+  return {
+    helpersHash: params.helpersHash,
+    permissions: params.permissions,
+    versionTag: params.pluginSetupRef.versionTag,
+    pluginAddress: params.plugin,
+    pluginRepo: params.pluginSetupRef.pluginSetupRepo,
+  };
+}
+
+export function getNamedTypesFromMetadata(
+  inputs: MetadataAbiInput[] = [],
+): string[] {
+  return inputs.map((input) => {
+    if (input.type.startsWith("tuple")) {
+      const tupleResult = getNamedTypesFromMetadata(input.components).join(
+        ", ",
+      );
+
+      let tupleString = `tuple(${tupleResult})`;
+
+      if (input.type.endsWith("[]")) {
```

Comment:
Would this work if something has a matrix? `[][]`

Reply:
I'm not sure if this can ever happen, because in the case of a tuple[] you would have components, and then you would have another tuple[] inside the components of this tuple and this recursive case is covered, but Im not sure if `tuple[][]` is a supported type. Since @Michael-A-Heuer wrote this maybe he can give us the response to this

Reply:
OS-513 created to tackle this